### PR TITLE
xyluo25 refactor from master: refactored utils.py and summary.py using Python 3.10, CI tests passed

### PR DESCRIPTION
**Pull request by xyluo25**
Dear Developers @martinfleis and @Ziqi-Li ,

I'm pleased to inform you that this Pull Request has successfully refactored **mgwr/utils.py** and **mgwr/summary.py** to be compatible with Python 3.10. I intend to further update additional documents, each through separate PRs for thorough validation. Please note that the current PR has also passed all CI tests.

Best regards,
Xiangyong Roy Luo

### CODE REVIEW

**Comment by xyluo25** [comment]
Hello Developers,

I wanted to let you know that I have reverted the previous merges to align with the most recent commit from the main branch under pysql/mgwr. This Pull Request specifically refactors utils.py and summary.py from the main branch. Thank you for your attention to this matter.

Additionally, after your validation, I will then submit new PR for merging into the main branch as well.

Best,
Xiangyong Roy Luo

File: mgwr/summary.py [hunk 1]
```diff
@@ -4,27 +4,32 @@
...
-        XNames = list(self.name_x)
-        if len(XNames) < self.k:
-            XNames = ["Intercept"] + XNames
-    else:
-        XNames = ["X" + str(i) for i in range(self.k)]
+    Returns:
+        str: print out the GLM summary information
+    """
+
+    XNames = [f"X{i}" for i in range(self.k)]
```
Comment (martinfleis) [inline mgwr/summary.py:27]: This is not equivalent of the original code. You are removing functionality here.

File: mgwr/summary.py [hunk 2]
```diff
@@ -52,20 +57,20 @@ def summaryGLM(self):
                                                 '-' * 10, '-' * 10)
     for i in range(self.k):
         summary += "%-31s %10.3f %10.3f %10.3f %10.3f\n" % (
-            XNames[i][:30], glm_rslt.params[i], glm_rslt.bse[i],
+            XNames[i], glm_rslt.params[i], glm_rslt.bse[i],
```
Comment (martinfleis) [inline mgwr/summary.py:60]: Why are you removing `[:30]`? That is not a stylistic change and there is a reason it is in the original code.

File: mgwr/summary.py [hunk 3]
```diff
@@ -52,20 +57,20 @@ def summaryGLM(self):
...
+def summaryGWR(self) -> str:
+    """Summary of GWR
 
-    else:
-        XNames = ["X" + str(i) for i in range(self.k)]
+    Returns:
+        str: print out the GWR summary information
+    """
+
+    XNames = ["X" + str(i) for i in range(self.k)]
```
Comment (martinfleis) [inline mgwr/summary.py:73]: Again, this is not equivalent!

File: mgwr/summary.py [hunk 4]
```diff
@@ -122,7 +127,7 @@ def summaryGWR(self):
         '-' * 20, '-' * 10, '-' * 10, '-' * 10, '-' * 10, '-' * 10)
     for i in range(self.k):
         summary += "%-20s %10.3f %10.3f %10.3f %10.3f %10.3f\n" % (
-            XNames[i][:20], np.mean(self.params[:, i]), np.std(self.params[:, i]),
```
Comment (martinfleis) [inline mgwr/summary.py:125]: Same as above.

File: mgwr/summary.py [hunk 5]
```diff
@@ -131,15 +136,14 @@ def summaryGWR(self):
...
-        XNames = list(self.name_x)
-        if len(XNames) < self.k:
-            XNames = ["Intercept"] + XNames
+    Returns:
+        str: print out the MGWR summary information
+    """
 
-    else:
-        XNames = ["X" + str(i) for i in range(self.k)]
+    XNames = ["X" + str(i) for i in range(self.k)]
```
Comment (martinfleis) [inline mgwr/summary.py:146]: Same as above

File: mgwr/summary.py [hunk 6]
```diff
@@ -199,9 +203,9 @@ def summaryMGWR(self):
         '-' * 20, '-' * 10, '-' * 10, '-' * 10, '-' * 10, '-' * 10)
     for i in range(self.k):
         summary += "%-20s %10.3f %10.3f %10.3f %10.3f %10.3f\n" % (
-            XNames[i][:20], np.mean(self.params[:, i]), np.std(self.params[:, i]),
+            XNames[i], np.mean(self.params[:, i]), np.std(self.params[:, i]),
```
Comment (martinfleis) [inline mgwr/summary.py:206]: Same as above

File: mgwr/utils.py [hunk 7]
```diff
@@ -1,34 +1,44 @@
...
-      should be set to  1 - 5/(5 + 15)) or 0.75
-    stop : Offset from highets point in the colormap's range.
-      Defaults to 1.0 (no upper ofset). Should be between
-      `midpoint` and 1.0.
-    
-    Returns
-    -------
-    new_cmap : A new colormap that has been shifted. 
-    '''
+def shift_colormap(cmap: LinearSegmentedColormap,
```
Comment (martinfleis) [inline mgwr/utils.py:8]: Are you sure that only `LinearSegmentedColormap` is allowed? not other types of Colormaps?

File: mgwr/utils.py [hunk 8]
```diff
@@ -1,34 +1,44 @@
...
+                   start: float = 0,
+                   midpoint: float = 0.5,
+                   stop: float = 1.0,
+                   name: str = 'shiftedcmap') -> LinearSegmentedColormap:
+
+    """Function to offset the "center" of a colormap.
+    Useful for data with a negative min and positive max
+    and you want the middle of the colormap's dynamic range to be at zero
+
+    Args:
```
Comment (martinfleis) [inline mgwr/utils.py:18]: This is not following numpydoc style while the original docstring does!

File: mgwr/utils.py [hunk 9]
```diff
@@ -59,61 +69,97 @@ def shift_colormap(cmap, start=0, midpoint=0.5, stop=1.0, name='shiftedcmap'):
...
-             bandwidth for mgwr model for var2
-    name   : string
-             common variable name to use for title
-    kwargs1:
-             additional plotting arguments for gwr surface
-    kwargs2:
-             additional plotting arguments for mgwr surface
-    savefig: string, optional
-             path to save the figure. Default is None. Not to save figure.
+    Args:
```
Comment (martinfleis) [inline mgwr/utils.py:121]: You are breaking numpydoc style again

**Review by martinfleis — CHANGES_REQUESTED** [review]
This does not follow what I explained in #144 how PRs should be structured. At the same time, it brings destructive changes to both code and its functionality and our docstrings. 

None of these changes is needed or agreed upon and at this point and quality only waste the maintainer time. I am sorry to be the bearer of bad news but this activity is not helping much.

File: mgwr/utils.py (hunk 9 above)
Comment (xyluo25) [reply mgwr/utils.py:121]: Hi Martin,

Thank you for your recommendation. I understand that Sphinx accepts both NumPyDoc and Google Docstring styles. Given Google Docstring's broader acceptance, I've been using it for new functions. From your comments, it seems you're advocating for consistency in existing documentation style. Acknowledging your point, I will also adhere to the NumPyDoc style to maintain this consistency.

File: mgwr/utils.py (hunk 7 above)
Comment (xyluo25) [reply mgwr/utils.py:8]: Hi @martinfleis,

I've reviewed the existing codebase and found that LinearSegmentedColormap is the only colormap utilized in MGWR. If I've overlooked any additional colormaps currently in use, your assistance in identifying them would be greatly appreciated.

**Pull request closed without merging**